crates/ra_hir/src/code_model.rs

```diff
@@ -196,14 +196,30 @@ impl ModuleDef {
             ModuleDef::Function(it) => Some(it.name(db)),
             ModuleDef::EnumVariant(it) => Some(it.name(db)),
             ModuleDef::TypeAlias(it) => Some(it.name(db)),
-
             ModuleDef::Module(it) => it.name(db),
             ModuleDef::Const(it) => it.name(db),
             ModuleDef::Static(it) => it.name(db),
 
             ModuleDef::BuiltinType(it) => Some(it.as_name()),
         }
     }
+
+    pub fn resolver<D: DefDatabase + HirDatabase>(&self, db: &D) -> Option<Resolver> {
```

Comment:
Resolver should not be exposed from `hir` crate, it's sort of the purpose of `code_model` to shield the caller from name resolution details. See https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/README.md#scale-of-changes for general philosophiy about this area

Reply:
As someone who hasn't looked at RA before (or anything compiler-like) I saw that `HasResolver` was implemented for `XxId` from `ra_hir_def` and that `ModuleDef` was just a wrapper around them. It seemed natural to do this.

What would you suggest doing here? I could just inline this method back into `try_resolve_intra` where it was earlier?

Reply:
>It seemed natural to do this.

Yeah, we are kind of still figuring internal structure, so it's not always obvious where the boundaries between layers are... At the moment this is only spelled in docs...

So I think that we should either have a method for each type in code_model (as  https://github.com/rust-analyzer/rust-analyzer/pull/4873#discussion_r463682352 suggested), or maybe only a single method on `SemanticsScope`. I think I'd lean towards first option, as it has a very narrow semantics: "resolve path from docs".

Hm, actually, I think this can be a method on `pub trait Docs`

Reply:
I ended up implementing it as part of a trait in `ide` which is used as a generic bound on the new `hir::link_rewrite::resolve_doc_link` API. The docs trait is implemented for a slightly different set of types.